hail/src/main/scala/is/hail/io/RowStore.scala

```diff
@@ -986,343 +935,224 @@ object EmitPackDecoder {
-            val wantIdx = fieldToWantIdx(fieldIdx)
-            val fieldSkip = skip || (wantIdx < 0)
-            val fieldType = t.types(fieldIdx)
-            val wantType = if (fieldSkip) fieldType else wantStruct.types(wantIdx)
-            val wantOffset = if (fieldSkip) -1 else wantStruct.byteOffsets(wantIdx)
-            if (!t.fieldRequired(fieldIdx)) {
-              val m = t.missingIdx(fieldIdx)
-              mainCode.append(s"if (!is_missing(${miss}, ${m})) {\n")
-              if (!fieldSkip) {
-                if (shuffleMissingBits) {
-                  val mbit = wantStruct.missingIdx(wantIdx)
-                  mainCode.append(s"  ${addr}[${mbit>>3}] &= ~(1<<${mbit&0x7});\n")
-                }
-              }
-              mainCode.append(s"  // ${name}.${field.name} fieldSkip ${fieldSkip} ${fieldType}\n")
-              scan(depth+1, s"${name}.${field.name}", fieldType, wantType, fieldSkip, baseAddr, off+wantOffset)
-              mainCode.append(s"}\n")
-            } else {
-              scan(depth+1, s"${name}.${field.name}", fieldType, wantType, fieldSkip, baseAddr, off+wantOffset)
-            }
-            fieldIdx += 1
-          }
+  def decodeBaseStruct(t: PBaseStruct, rt: PBaseStruct, input_buf_ptr: cxx.Expression, region: cxx.Expression, off: cxx.Expression): cxx.Code = {
+    val ssb = new cxx.StagedBaseStructBuilder(rt, off)
+    def wrapMissing(missingBytes: cxx.Code, tidx: Int)(processField: cxx.Code): cxx.Code = {
+      if (!t.fieldRequired(tidx)) {
+        s"""if (!${t.cxxIsFieldMissing(missingBytes, tidx)}) {
+           |  $processField
+           |}""".stripMargin
+      } else processField
+    }
+    if (t.size == rt.size) {
+      assert((t.isInstanceOf[PTuple] && rt.isInstanceOf[PTuple]) || (t.asInstanceOf[PStruct].fieldNames sameElements t.asInstanceOf[PStruct].fieldNames))
+      val decodeFields = Array.tabulate[cxx.Code](rt.size) { idx =>
+        wrapMissing(s"$off", idx)(ssb.addField(idx, foff => decode(t.types(idx), rt.types(idx), input_buf_ptr, region, cxx.Expression(foff))))
+      }.mkString("\n")
+      if (t.nMissingBytes > 0) {
+        s"""$input_buf_ptr->read_bytes($off, ${ t.nMissingBytes });
+           |$decodeFields""".stripMargin
+      } else
+        decodeFields
+    } else {
+      val names = t.asInstanceOf[PStruct].fieldNames
+      val rnames = rt.asInstanceOf[PStruct].fieldNames
+      val missing_bytes = cxx.ArrayVariable(s"missing", "char", s"${t.nMissingBytes}")
+
+      var rtidx = 0
+      val decodeFields = Array.tabulate[cxx.Code](t.size) { tidx =>
+        val f = t.types(tidx)
+        val skipField = rtidx >= rt.size || names(tidx) != rnames(rtidx)
+        val processField = if (skipField)
+          skip(f, input_buf_ptr)
+        else
+          s"""${ if (f.required) "" else ssb.clearMissing(rtidx) }
+             |${ ssb.addField(rtidx, foff => decode(f, rt.types(rtidx), input_buf_ptr, region, cxx.Expression(foff))) }"""
```

Comment:
Another style question, should the first line be with the """ or on a line with a |?  Once we have the bar, I think it's more readable to have every line start with a bar.